### PR TITLE
GH-112727: Speed up `pathlib.Path.absolute()`

### PR DESCRIPTION
**Pull request by barneygale**
Use `_from_parsed_parts()` to create a pre-joined/pre-parsed path, rather than passing multiple arguments to `with_segments()`

* Issue: gh-112727

### CODE REVIEW

**Comment by barneygale** [comment]
1.5x speedup for paths with tail segments:

```
$ ./python -m timeit -s "from pathlib import Path" "str(Path().absolute())"
100000 loops, best of 5: 3.82 usec per loop  # before
100000 loops, best of 5: 3.82 usec per loop  # after

$ ./python -m timeit -s "from pathlib import Path" "str(Path('foo').absolute())"
50000 loops, best of 5: 9.79 usec per loop  # before
50000 loops, best of 5: 6.14 usec per loop  # after

$ ./python -m timeit -s "from pathlib import Path" "str(Path('foo', 'bar').absolute())"
20000 loops, best of 5: 11.5 usec per loop  # before
50000 loops, best of 5: 7.49 usec per loop  # after
```

It would be interesting to also test the performance of `str(WindowsPath('c:foo'))` and `str(WindowsPath('/foo'))`, but I don't have a Windows box to hand.

**Comment by AlexWaygood** [comment]
Big speedups on Windows!

On `main`:

```
>python -m timeit -s "from pathlib import WindowsPath" "str(WindowsPath('c:foo').absolute())"
Running PGUpdate|x64 interpreter...
50000 loops, best of 5: 9.92 usec per loop

>python -m timeit -s "from pathlib import WindowsPath" "str(WindowsPath('/foo').absolute())"
Running PGUpdate|x64 interpreter...
50000 loops, best of 5: 8.32 usec per loop
```

With PR:

```
>python -m timeit -s "from pathlib import WindowsPath" "str(WindowsPath('c:foo').absolute())"
Running PGUpdate|x64 interpreter...
50000 loops, best of 5: 6.01 usec per loop

>python -m timeit -s "from pathlib import WindowsPath" "str(WindowsPath('/foo').absolute())"
Running PGUpdate|x64 interpreter...
50000 loops, best of 5: 4.97 usec per loop
```

(Timings done using a reasonably fresh PGO-optimised build)

**Comment by AlexWaygood** [comment]
I can also repro all your reported speedups from https://github.com/python/cpython/pull/112728#issue-2024809845!

File: Lib/pathlib.py [hunk 1]
```diff
@@ -1415,21 +1415,29 @@ def absolute(self):
         """
         if self.is_absolute():
             return self
+        elif self.root:
+            drive = os.path.splitroot(os.getcwd())[0]
+            return self._from_parsed_parts(drive, self.root, self._tail)
         elif self.drive:
```
Comment (AlexWaygood) [inline Lib/pathlib.py]: I'd find this more readable. The `cwd` variable is really set on one of two branches: line 1423 and line 1425. It makes sense for those two branches to be a separate `if/else` suite, after the first two early-return opportunites, I think:

```suggestion
        if self.root:
            drive = os.path.splitroot(os.getcwd())[0]
            return self._from_parsed_parts(drive, self.root, self._tail)
        if self.drive:
```

**Review by AlexWaygood — APPROVED** [review]
LGTM

**Comment by barneygale** [comment]
Thanks tons!

**Pull request merged**